### ISSUE TITLE
[App]: define_metrics does not work with media type keys

### ISSUE THREAD

**Issue by aria1th** [wandb/wandb#6500]
### Current Behavior

Assume the following code:
```
        wandb.define_metric("custom_step")
        wandb.define_metric("image_0", step_metric='custom_step')
        wandb.log(
            {
                'custom_step' : steps,
                'image_0': wandb.Image(image),
            }
        )
```

Then wandb can't show the custom metric vs media type plot, it will show with default step (current wandb step) instead.

### Expected Behavior

```
        wandb.define_metric("custom_step")
        wandb.define_metric("metric_0", step_metric='custom_step')
        wandb.log(
            {
                'custom_step' : steps,
                'metric_0' : 0,
            }
        )
```

this works well.

### Steps To Reproduce

Define the custom metric with the code above, then try to log the images, to see plotted image against custom metrics.

### Environment

OS: Linux / Windows *all

Browsers: Chrome / Microsoft Edge

Version: 0.15.12

### Additional Context

Related : https://github.com/wandb/wandb/issues/6077

**Comment by sydholl** [comment]
Internal GitHub User commented: 
### Current Behavior

Assume the following code:
```
        wandb.define_metric("custom_step")
        wandb.define_metric("image_0", step_metric='custom_step')
        wandb.log(
            {
                'custom_step' : steps,
                'image_0': wandb.Image(image),
            }
        )
```

Then wandb can't show the custom metric vs media type plot, it will show with default step (current wandb step) instead.

### Expected Behavior

```
        wandb.define_metric("custom_step")
        wandb.define_metric("metric_0", step_metric='custom_step')
        wandb.log(
            {
                'custom_step' : steps,
                'metric_0' : 0,
            }
        )
```


this works well.

### Steps To Reproduce

Define the custom metric with the code above, then try to log the images, to see plotted image against custom metrics.

### Screenshots

_No response_

### Environment

OS: Linux / Windows *all

Browsers: Chrome / Microsoft Edge

Version: 0.15.12


### Additional Context

Related : https://github.com/wandb/wandb/issues/6077

**Comment by JoanaMarieL** [comment]
Hi @aria1th ,

Thank you for writing in, we will review the codes you provided and will get back to you.

**Comment by JoanaMarieL** [comment]
Hello @aria1th ,

Yes, it's true. Weights & Biases uses the internal step counter as the default x-axis for your plots. If you want to use a custom metric as the x-axis, you can use the wandb.define_metric function to specify the metric you want to use as the x-axis. However, this only applies to scalar metrics and not media types.

We can also create a feature request for this case if you would like.

**Comment by aria1th** [comment]
@JoanaMarieL Thank you for the reply!
 Sure, the feature would be great. We can think about some asynchronous inference while training is running background. 

(Also, inference can go slow, so it may end after actual training is done... but currently if we end the training, there is no way to attach media results. Are there way to bypass this?)

**Comment by JoanaMarieL** [comment]
Hi @aria1th ,

I would like to clarify the last statement, do you mean to attach the media results after the actual training end?

**Comment by aria1th** [comment]
@JoanaMarieL Yes, I once tried to attach results after end of training - for example, I assigned the thread to call the inference, but since `accelerator.end_training()` was called before inference end the stuff, it was not able to attach results.

Of course the best solution is to not call end_training if inference is not done, but if some bypass behaviors / queue / reservation is available, it would be far convenient.

**Comment by JoanaMarieL** [comment]
@aria1th , can you give this a try:

by resuming run:  [https://docs.wandb.ai/guides/runs/resuming#resume-guidance]
You can resume then add media results or upload them as artifacts to the same run.
It works like this:
start a run and log
```
run = wandb.init(entity, project)
run.log(...)
run_id = run.id
run.finish()
```
2. resume the run run_id anytime and log more data
```
run = wandb.init(entity, project, id= run_id, resume='must')
run.log(..)
run.finish()
```

**Comment by aria1th** [comment]
@JoanaMarieL I'm using the following code in accelerator environment, but seems there is some problem:
```
        global run_id
        if run_id is None:
            run_id = wandb.run.id
        # resume if stopped
        wandb.init(id=run_id, resume="must")
```

This function is called every inference. Somehow, the first inference (with run didn't finish) worked, but later ones didn't work - it had problem while waiting wandb.init(id=run_id)

Actually, it failed to resume the run as I checked -

![image](https://github.com/wandb/wandb/assets/35677394/94874919-b71e-4974-9b1c-97aca8759403)

I tested if resume works with "allow", "must", but seems like it didn't work, it has thrown error after 90 seconds.

![image](https://github.com/wandb/wandb/assets/35677394/c6415346-9553-4271-b332-dbd3540d7f84)

Full code snippet:

```
import wandb
wandb.login(key=wandb_api_token)

# 2. Start a new run
wandb.init(project='test')

run_id = wandb.run.id

run = wandb.run

run.log({'acc': 0.9, 'loss': 0.2}) #test

run.finish()

run = wandb.init(id=run_id, resume="allow") # throws CommError
```

**Comment by JoanaMarieL** [comment]
@aria1th , let us review the information provided and will get back to you.

**Comment by JoanaMarieL** [comment]
Hi @aria1th ,
please try to modify this part by adding entity and project like this: [`wandb.init(entity="entity_name",project='proj',name="test")`]
```
# 2. Start a new run
wandb.init(project='test')
```
let us know the result after you try again

**Comment by aria1th** [comment]
@JoanaMarieL Thanks for the response!
I successfully resumed the run and logged the metrics with the following code snippet:
```python
import wandb
wandb.login(key=API_KEY)

wandb.init(entity="user_name",project='proj_test',name="test_name")

wandb.log({'acc': 0.9, 'loss': 0.2}) #test

args = run.project, run.entity, run.name, run.id

run.finish()

# Note, this may interrupt main training process run - so we need to check 'finished' before calling this

run = wandb.init(resume="must", project=args[0], entity=args[1], name=args[2], id= args[3])

run.log({'acc': 0.9, 'loss': 0.2}) #test

run.finish()
```

Unfortunately, media types are not available with custom steps - which is the only problem.

**Comment by JoanaMarieL** [comment]
Hello @aria1th 
We're glad that this works now for you. And as for the custom steps using Media Types, we will file this as a Feature Request. For now we will mark this as resolved. Feel free to reach us out again for any concerns.

**Comment by vmichals** [comment]
@JoanaMarieL Thank you for the information!
Custom steps for media types would indeed be very useful!
In the meantime it'd be great to add a note in the documentation, indicating that the `step_metric`  argument of `define_metric` only works for scalar type metrics. I ran into this issue and was investigating whether the run was maybe not syncing properly.

**Comment by JoanaMarieL** [comment]
Thank you for flagging @vmichals , we'll definitely include this to our documentation.

**Comment by heilrahc** [comment]
So does media type supports custom x-axis now? When I try it it still can't, the media data is always logged to the current step even when I explicitly defined a custom x-axis. This is a very important feature that should be implemented.

**Comment by JoanaMarieL** [comment]
hi @heilrahc , this feature is still in queue on our end. we will provide an update once there is any progress.

**Comment by jnzhnd** [comment]
Any update on this?
